### PR TITLE
Generate: FLAX infers pad token in its absence and has functional example

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Some bug fixing in advance of #21007 (PR that adds generation config to Flax), to ensure we start from a functional flax generate codebase. 

In particular:
1. Flax now assumes the value `pad_token_id` when it is `None` and `eos_token_id` is not `None`, like TF and PT do. This is very helpful for open text generation examples, like with GPT2, was an open request (https://github.com/huggingface/transformers/issues/18884), and was one of the causes for failure in the existing example. This also includes the recent changes of #20727, where `eos_token_id` can be a list of tokens.
2. An `int32` type specification was missing in the special tokens -- when converted to JAX variables, JAX assumed they were `float32`;
3. The existing flax generate example is now part of our doctests, and runs.

### CODE REVIEW

File: src/transformers/generation/tf_utils.py [hunk 1]
```diff
@@ -702,11 +702,11 @@ def generate(
...
-            logger.warning(
-                f"Setting `pad_token_id` to {generation_config.eos_token_id} (first `eos_token_id`) to generate"
-                " sequence"
-            )
-            generation_config.pad_token_id = generation_config.eos_token_id
+            eos_token_id = generation_config.eos_token_id
+            if isinstance(eos_token_id, list):
+                eos_token_id = eos_token_id[0]
+            logger.warning(f"Setting `pad_token_id` to `eos_token_id`:{eos_token_id} for open-end generation.")
+            generation_config.pad_token_id = eos_token_id
```
Comment (gante) [inline src/transformers/generation/tf_utils.py:709]: Took the opportunity also to copy the logic to TF, so it can also handle `eos_token_id` as a list 👀

**Review by sgugger — APPROVED** [review]
Thanks for the fix!

**Review by patrickvonplaten — APPROVED** [review]
Thanks!

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks for the fix @gante!

**Pull request merged**